usr/sbin/newfs/newfs.8
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)newfs.8	2.0 (2.11BSD) 1996/04/12
.\"
.TH NEWFS 8 "April 12, 1996"
.UC 5
.SH NAME
newfs \- construct a new file system
.SH SYNOPSIS
.B /sbin/newfs
[
.B \-N
]
[
.B \-m free-gap
]
[
.B \-n free-modulus
]
[
.B \-i bytes
]
[
.B \-s size
]
[
.B \-T disk-type
]
.B special
.SH DESCRIPTION
.B Newfs
is a ``friendly'' front-end to the
.BR mkfs (8)
program.
.BR Newfs (8)
will normally read the disklabel from the drive to determine the 
partition sizes.  If the driver for the disk does not support 
disklabels the \-T option must be used to force a search of
.IR /etc/disktab
for partition information about \fIdrive-type\fP.
.B Newfs
calculates the appropriate parameters to use in calling
.BR mkfs ,
then builds the file system by forking \fBmkfs\fP.
.PP
.B \-N
causes the \fBmkfs\fP command which would be executed to be printed out
without actually creating the file system.  The disk specified by \fBspecial\fP
must be online though so that \fBnewfs\fP can read the disklabel.
.PP
.B \-m
allows the specification of the block interleaving of the free list.
If not specified or outside the range 1 thru 32 then 
a value of 2 is used.
.PP
.B \-n
parameter is the freelist modulus (when the \fB\-m\fP pattern repeats)
and is calculated by
.B newfs
to be 1 cylinder in size by default.
.PP
.B \-i
specifies how many bytes per inode to assume when calculating how many
inodes to allocate.  The default is 4096 bytes per inode.  If this results
in too few inodes being allocated (there is an absolute maximum of 65500)
then decrease the \fIbytes\fP number (which must lie between 512 and 65536).
.PP
.B \-T
must be used if the disk specified by \fIspecial\fP has not
been labeled with the
.BR disklabel (8)
program.  In this case \fIdisk-type\fP is used by 
.BR getdisklabel (3)
when searching
.IR /etc/disktab .
This option is used when the underlying device driver does not
support disklabels.  Care must be taken that the contents of
/etc/disktab match the partition tables in the kernel.
.PP
.BR \-s
specifies how many \fBsectors\fP the file system is to contain.
There are two sectors per file system block, therefore \fIsize\fP should
be even.
This parameter must be less than or equal to the partition size (as 
determined from the disklabel or /etc/disktab).  An error is printed and
no action is taken if the partition size is 0 or too large.
.PP
.in +0.5i
.B NOTE:
Mkfs deals in units of filesystem blocks not sectors.  Newfs uses sectors.
.in -0.5i
.SH FILES
/etc/disktab	disk geometry and partition information 
.br
mkfs	to actually build the file system
.SH "SEE ALSO"
getdisklabel(3), disklabel(8), disktab(5), diskpart(8), fs(5), fsck(8),
mkfs(8)
.SH BUGS
.BR newfs (8)
no longer places boot blocks on the filesystem.  That duty has been 
moved to the 
.BR disklabel (8)
program.  If you must place a boot block on a disk whose driver does
not support disklabels use 
.BR dd (1).


usr/sbin/newfs/newfs.c
#define	COMPAT

/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#if	!defined(lint) && defined(DOSCCS)
char copyright[] =
"@(#) Copyright (c) 1983 Regents of the University of California.\n\
 All rights reserved.\n";

static char sccsid[] = "@(#)newfs.c	6.3 (2.11BSD) 1996/11/16";
#endif

/*
 * newfs: friendly front end to mkfs
 *
 *  The installation of bootblocks has been moved to 'disklabel(8)'.  This
 *  program expects to find a disklabel present which will contain the
 *  geometry and partition size information.  If 'newfs' is being run on
 *  a system without disklabels implemented in the kernel you must specify
 *  "-T diskname" (where 'diskname' is an entry in /etc/disktab).
 */

#include <stdio.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/disklabel.h>
#include <sys/disk.h>
#include <sys/file.h>

#include <ctype.h>
#include <errno.h>
#include <paths.h>
#include <stdlib.h>
#include <syslog.h>
#include <varargs.h>

#ifdef	COMPAT
	char	*disktype;
	int	unlabeled;
#endif
	int	Nflag;
	struct	disklabel *getdisklabel();

extern	char	*__progname;

main(argc, argv)
	int	argc;
	char	**argv;
{
	register struct disklabel	*lp;
	register struct partition	*pp;
	char	*cp;
	struct stat	st;
	long	fssize, ltmp;
	int	f_n = 0, f_m = 0;
	u_int	f_i = 4096;
	int	ch, status, logsec;
	int	fsi;
	char	device[MAXPATHLEN], cmd[BUFSIZ], *index(), *rindex();
	char	*special;

	while ((ch = getopt(argc,argv,"T:Nvm:s:n:i:")) != EOF)
		switch((char)ch) {
		case 'N':
		case 'v':
			Nflag = 1;
			break;
#ifdef	COMPAT
		case 'T':
			disktype = optarg;
			break;
#endif
		case 'm':
			ltmp = atol(optarg);
			if	(ltmp <= 0 || ltmp > 32)
				fatal("%s: out of 1 - 32 range", optarg);
			f_m = (int)ltmp;
			break;
		case	'n':
			ltmp = atol(optarg);
/*
 * If the upper bound is changed here then mkfs.c must also be changed
 * also else mkfs will cap the value to its limit.
*/
			if	(ltmp <= 0 || ltmp > 500)
				fatal("%s: out of 1 - 500 range", optarg);
			f_n = (int)ltmp;
			break;
		case	'i':
			ltmp = atol(optarg);
			if	(ltmp < 512 || ltmp > 65536L)
				fatal("%s: out of 512 - 65536 range", optarg);
			f_i = (u_int)ltmp;
			break;
		case 's':
			fssize = atol(optarg);
			if (fssize <= 0)
				fatal("%s: bad file system size", optarg);
			break;
		case '?':
		default:
			usage();
		}

	argc -= optind;
	argv += optind;
	if (argc != 2 && argc != 1)
		usage();

	/* figure out device name */
	special = argv[0];
	cp = rindex(special, '/');
	if (cp == 0) {
		/*
		 * No path prefix; try /dev/r%s then /dev/%s
		 */
		 (void)sprintf(device, "%s/r%s", _PATH_DEV, special);
		 if (stat(device, &st) == -1)
			(void)sprintf(device, "%s/%s", _PATH_DEV, special);
		special = device;
	}

	fsi = open(special, O_RDONLY);
	if (fsi < 0)
		fatal("%s: %s", special, strerror(errno));

	/* see if it exists and of a legal type */
	if (fstat(fsi, &st) == -1)
		fatal("%s: %s", special, strerror(errno));
	if ((st.st_mode & S_IFMT) != S_IFCHR)
		fatal("%s: not a character device", special);
	cp = index(argv[0], '\0') - 1;
	if (cp == 0 || (*cp < 'a' || *cp > 'h') && !isdigit(*cp))
		fatal("%s: can't figure out file system partition", argv[0]);

#ifdef	COMPAT
	if (disktype == NULL)
		disktype = argv[1];
#endif
	lp = getdisklabel(special, fsi);
	if (isdigit(*cp))
		pp = &lp->d_partitions[0];
	else
		pp = &lp->d_partitions[*cp - 'a'];
	if (pp->p_size <= 0)
		fatal("%s: '%c' partition is unavailable", argv[0], *cp);
#ifdef	nothere
	if (pp->p_fstype == FS_BOOT)
		fatal("%s: '%c' partition  overlaps boot program",argv[0], *cp);
#endif

	if (fssize == 0)
		fssize = pp->p_size;
	if (fssize > pp->p_size)
		fatal("%s: maximum file system size on '%c' partition is %ld",
			argv[0], *cp, pp->p_size);
	/*
	 * Convert from sectors to logical blocks.  Note that sector size
	 * must evenly devide DEV_BSIZE!!!!!
	 */

	/*
	 * getdisklabel(3) forces the sector size to 512 because any other
	 * choice would wreck havoc in the disklabel(8) program and result
	 * in corrupted/destroyed filesystems.  DEV_BSIZE had better be
	 * 1024!
	*/
	if (lp->d_secsize != 512)
		fatal("%s: sector size not 512", argv[0]);
	logsec = DEV_BSIZE/lp->d_secsize;
	fssize /= logsec;

	/* build command */
	if	(f_m == 0)	/* If never specified then use default of 2 */
		f_m = 2;
	if	(f_n == 0)	/* If never specified then 1/2 the cyl size */
		f_n = lp->d_secpercyl / logsec;

	sprintf(cmd, "/sbin/mkfs -m %d -n %d -i %u -s %ld %s", f_m, f_n, f_i,
		fssize, special);
	printf("newfs: %s\n", cmd);

	close(fsi);

	if (Nflag)
		exit(0);

	if (status = system(cmd))
		exit(status >> 8);
	exit(0);
}

#ifdef COMPAT
char lmsg[] = "%s: can't read disk label; disk type must be specified";
#else
char lmsg[] = "%s: can't read disk label";
#endif

struct disklabel *
getdisklabel(s, fd)
	char *s;
	int fd;
{
	static struct disklabel lab;

	if (ioctl(fd, DIOCGDINFO, (char *)&lab) < 0) {
#ifdef COMPAT
		if (disktype) {
			struct disklabel *lp, *getdiskbyname();

			unlabeled++;
			lp = getdiskbyname(disktype);
			if (lp == NULL)
				fatal("%s: unknown disk type", disktype);
			return (lp);
		}
#endif
		warn("ioctl (GDINFO)");
		fatal(lmsg, s);
	}
	return (&lab);
}

static
usage()
{

	fprintf(stderr,"usage: %s [-N] [-m freelist-gap] [-s filesystem size] ",
		__progname);
	fprintf(stderr, "[-i bytes/inode] [-n freelist-modulus] ");
#ifdef	COMPAT
	fputs("[-T disk-type] ", stderr);
#endif
	fputs("special-device\n", stderr);
	exit(1);
}

/*VARARGS*/
void
fatal(fmt, va_alist)
	char *fmt;
	va_dcl
{
	va_list ap;

	va_start(ap);

	if (fcntl(fileno(stderr), F_GETFL) < 0) {
		openlog(__progname, LOG_CONS, LOG_DAEMON);
		vsyslog(LOG_ERR, fmt, ap);
		closelog();
	} else {
		vwarnx(fmt, ap);
	}
	va_end(ap);
	exit(1);
	/*NOTREACHED*/
}
